### PR TITLE
Sprint1/tictactoe/add position to api

### PR DESCRIPTION
**Pull request by martingaston**
This PR once again updates the `tictactoe` API package. 

This addresses an oversight on my part of how the incoming/outgoing JSON objects are likely to look: 

- incoming JSON will feature a `position` key with the value of the index of the move to be played. In my previous designs there was an `update` method that took two arguments, but now I realise that the stringified JSON object needs to contain pretty much everything. Whoops! 
- outgoing JSON does not feature said `position` field. 

This creates an interesting scenario in a statically typed language where, well, I've ended up making two separate objects to accommodate for this - `JsonIncoming` and `JsonOutgoing`, along with a `Json` interface to declare the fields required in *all* requests. 

This has also put me in a position where the tests are, well, I think they're more brittle. I'm not really coming at them from the perspective of the public API, because I can't figure out a straightforward way to. Right now I've got a `Builder` class in `JsonIncoming` that, well, only exists for the tests, which feels like a smell as well. I'm not quite sure *how* to go about making them better. So I'm a little stumped here - it would be super to talk about this in a code review/the IPM 😎

### CODE REVIEW

File: tictactoe/src/main/java/com/github/martingaston/tictactoe/api/API.java [hunk 1]
```diff
@@ -11,9 +11,9 @@ public static String init(String mode) throws IOException {
         return Encode.from(Response.initial(mode));
     }
 
-    public static String update(int position, String jsonString) throws IOException {
-        GameJSON previousGame = Decode.from(jsonString);
-        GameJSON updated = Update.from(position, previousGame);
+    public static String update(String jsonString) throws IOException {
```
Comment (fabientownsend) [inline tictactoe/src/main/java/com/github/martingaston/tictactoe/api/API.java]: `API.update(string jsonString)` your public API sounds really generic as, it doesn't really communicate what it does, which is really important it is a public API.

`jsonString` is quire redundant as a name as your type specify that's the json is a string

Isn't `previousGame`, the current one, as it's not updated yet?

Comment (martingaston) [reply tictactoe/src/main/java/com/github/martingaston/tictactoe/api/API.java]: That is very true!

File: tictactoe/src/main/java/com/github/martingaston/tictactoe/api/Encode.java [hunk 2]
```diff
@@ -4,10 +4,10 @@
 import com.fasterxml.jackson.databind.ObjectMapper;
 
 class Encode {
-    static String from(GameJSON gameJson) throws JsonProcessingException {
+    static String from(JsonOutgoing jsonOutgoing) throws JsonProcessingException {
         var mapper = new ObjectMapper();
         mapper.configure(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY, true);
-        return mapper.writeValueAsString(gameJson);
+        return mapper.writeValueAsString(jsonOutgoing);
```
Comment (fabientownsend) [inline tictactoe/src/main/java/com/github/martingaston/tictactoe/api/Encode.java:10]: I feel like that the two classes could be one:

``` java
class GameJson {
   public Game objectify(Json game) {}
   pubic Json parse(Game game) {}
}
```

File: tictactoe/src/main/java/com/github/martingaston/tictactoe/api/Referee.java [hunk 3]
```diff
@@ -10,16 +10,16 @@
 import java.util.stream.Collectors;
 
 class Referee {
-    static Symbol swapPlayer(Symbol currentPlayer) {
-        return currentPlayer.equals(new Symbol("X")) ? new Symbol("O") : new Symbol("X");
-    }
+    static Symbol swapPlayer(JsonIncoming previousMove) {
+        if(previousMove.mode().equals("ai")) {
+            return new Symbol("X");
```
Comment (fabientownsend) [inline tictactoe/src/main/java/com/github/martingaston/tictactoe/api/Referee.java:15]: Is that part of the web API?

**Comment by fabientownsend** [comment]
I started to review it,  it will be easy to talk about it over during the IPM

**Pull request merged**